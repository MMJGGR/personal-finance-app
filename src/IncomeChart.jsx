import React, { useState, useMemo } from 'react'
import {
  BarChart,
  Bar,
  XAxis,
  YAxis,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from 'recharts'
import { useFinance } from './FinanceContext'
import { formatCurrency } from './utils/formatters'

export default function IncomeChart() {
  const { startYear, years, annualIncome, annualIncomePV, settings } = useFinance()
  const [mode, setMode] = useState('nominal')
  const data = useMemo(() => {
    const vals = mode === 'nominal' ? annualIncome : annualIncomePV
    return vals.map((v, idx) => ({
      year: String(startYear + idx),
      income: Number(v.toFixed(2)),
    }))
  }, [mode, annualIncome, annualIncomePV, startYear])
  const format = value => formatCurrency(value, settings.locale, settings.currency)
  return (
    <section className="bg-white p-4 rounded-xl shadow-md h-80">
      <div className="mb-2">
        <button
          onClick={() => setMode('nominal')}
          className={`px-3 py-1 rounded-full text-sm mr-2 ${
            mode === 'nominal' ? 'bg-amber-400 text-white' : 'bg-slate-200 text-slate-700'
          }`}
        >
          Nominal
        </button>
        <button
          onClick={() => setMode('discounted')}
          className={`px-3 py-1 rounded-full text-sm ${
            mode === 'discounted' ? 'bg-amber-400 text-white' : 'bg-slate-200 text-slate-700'
          }`}
        >
          Discounted
        </button>
      </div>
      <ResponsiveContainer width="100%" height="100%" role="img" aria-label="Income chart">
        <BarChart data={data}>
          <XAxis dataKey="year" />
          <YAxis />
          <Tooltip formatter={format} />
          <Legend />
          <Bar dataKey="income" fill="#f59e0b" name="Income" />
        </BarChart>
      </ResponsiveContainer>
    </section>
  )
}
