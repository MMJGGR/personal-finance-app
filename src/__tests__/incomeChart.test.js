import React, { useEffect } from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { FinanceProvider, useFinance } from '../FinanceContext'
import IncomeChart from '../IncomeChart'

beforeAll(() => {
  global.ResizeObserver = class { observe() {} unobserve() {} disconnect() {} }
})

afterEach(() => {
  localStorage.clear()
})

function Wrapper({ children }) {
  const { setYears, updateSettings, settings } = useFinance()
  useEffect(() => {
    setYears(2)
    updateSettings({ ...settings, discountRate: 10 })
  }, [])
  return children
}

test('discounted chart snapshot', () => {
  localStorage.setItem('incomeSources', JSON.stringify([
    { name: 'Job', amount: 1000, frequency: 1, growth: 0, taxRate: 0 }
  ]))
  const { container } = render(
    <FinanceProvider>
      <Wrapper>
        <IncomeChart />
      </Wrapper>
    </FinanceProvider>
  )
  fireEvent.click(screen.getByRole('button', { name: /Discounted/i }))
  expect(container.firstChild).toMatchSnapshot()
})

test('toggle switches chart data', () => {
  localStorage.setItem('incomeSources', JSON.stringify([
    { name: 'Job', amount: 1000, frequency: 1, growth: 0, taxRate: 0 }
  ]))
  render(
    <FinanceProvider>
      <Wrapper>
        <IncomeChart />
      </Wrapper>
    </FinanceProvider>
  )
  const nominal = screen.getByRole('button', { name: /Nominal/i })
  const disc = screen.getByRole('button', { name: /Discounted/i })
  expect(nominal).toHaveClass('bg-amber-400')
  fireEvent.click(disc)
  expect(disc).toHaveClass('bg-amber-400')
})
