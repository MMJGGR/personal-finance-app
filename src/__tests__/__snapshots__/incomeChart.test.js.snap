// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`discounted chart snapshot 1`] = `
<section
  class="bg-white p-4 rounded-xl shadow-md h-80"
>
  <div
    class="mb-2"
  >
    <button
      class="px-3 py-1 rounded-full text-sm mr-2 bg-slate-200 text-slate-700"
    >
      Nominal
    </button>
    <button
      class="px-3 py-1 rounded-full text-sm bg-amber-600 text-white"
    >
      Discounted
    </button>
  </div>
  <div
    class="recharts-responsive-container"
    style="width: 100%; height: 100%; min-width: 0;"
  />
</section>
`;
